File: cornelsen.js
#!/usr/bin/env node
/**
 * Extract books from Cornelsen Offline Lernen data.
 *
 * Reads the Cornelsen app container, decrypts the embedded PDF,
 * extracts page text via pdf.js, and exports supplementary materials
 * (HTML tips, additional PDFs).
 *
 * Usage: node bin/cornelsen [--output <dir>] [--book <id>]
 *                           [--no-materials] [--markdown] [--force]
 */

import { readFileSync, writeFileSync, existsSync, readdirSync, mkdirSync, rmSync } from "node:fs";
import { join, resolve, basename, dirname } from "node:path";
import { homedir, tmpdir } from "node:os";
import { createDecipheriv } from "node:crypto";
import { inflateRawSync } from "node:zlib";
import { fileURLToPath } from "node:url";

// -- pdf.js for text extraction (no system dependency) --
let pdfjsLib = null;
try {
  const __dirname = dirname(fileURLToPath(import.meta.url));
  const pdfjsPath = join(__dirname, "..", "node_modules", "pdfjs-dist", "legacy", "build", "pdf.mjs");
  if (existsSync(pdfjsPath)) {
    pdfjsLib = await import(pdfjsPath);
  }
} catch { /* pdfjs-dist not available */ }

// -- AES-128-CBC decryption (key assembled from app source) --
const CIPHER_B64 = "YWVzLTEyOC1jYmN8RCtEeEpTRn0yQjtrLTtDfQ==";

function decryptBuffer(encrypted) {
  const decoded = Buffer.from(CIPHER_B64, "base64").toString("ascii");
  const [algorithm, key] = decoded.split("|");
  const iv = key.split("").reverse().join("");
  const decipher = createDecipheriv(algorithm, key, iv);
  return Buffer.concat([decipher.update(encrypted), decipher.final()]);
}

// -- Path to Cornelsen app data --
function findCornelsenBase() {
  const candidates = [];

  if (process.platform === "darwin") {
    candidates.push(
      "/Applications/CornelsenOfflineLernen.app/Contents/Resources/uma",
      join(homedir(), "Applications", "CornelsenOfflineLernen.app",
        "Contents", "Resources", "uma"),
    );
  } else if (process.platform === "win32") {
    const localApp = process.env.LOCALAPPDATA || join(homedir(), "AppData", "Local");
    const progFiles = process.env.PROGRAMFILES || "C:\\Program Files";
    const progFiles86 = process.env["PROGRAMFILES(X86)"] || "C:\\Program Files (x86)";

    for (const base of [join(localApp, "Programs"), progFiles, progFiles86]) {
      // Download-Build: resources/uma/
      candidates.push(join(base, "CornelsenOfflineLernen", "resources", "uma"));
      // MSI-Build: uma/ direkt im App-Verzeichnis
      candidates.push(join(base, "CornelsenOfflineLernen", "uma"));
      // Alternativer Name mit Leerzeichen
      candidates.push(join(base, "Cornelsen Offline Lernen", "resources", "uma"));
      candidates.push(join(base, "Cornelsen Offline Lernen", "uma"));
    }
  }

  for (const p of candidates) {
    if (existsSync(p)) return p;
  }
  return candidates[0];
}

// -- Pure Node.js ZIP reader (no external tools) --

function readZipCentralDirectory(zipPath) {
  const buf = readFileSync(zipPath);

  // Find End of Central Directory (scan backwards for signature 0x06054b50)
  let eocdOffset = -1;
  for (let i = buf.length - 22; i >= Math.max(0, buf.length - 65557); i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) { eocdOffset = i; break; }
  }
  if (eocdOffset < 0) return [];

  let cdOffset = buf.readUInt32LE(eocdOffset + 16);
  const cdSize = buf.readUInt32LE(eocdOffset + 12);

  // Handle ZIP64: if cdOffset is 0xFFFFFFFF, look for ZIP64 EOCD locator
  if (cdOffset === 0xFFFFFFFF) {
    for (let i = eocdOffset - 20; i >= Math.max(0, eocdOffset - 40); i--) {
      if (buf.readUInt32LE(i) === 0x07064b50) {
        const zip64EocdOffset = Number(buf.readBigUInt64LE(i + 8));
        if (buf.readUInt32LE(zip64EocdOffset) === 0x06064b50) {
          cdOffset = Number(buf.readBigUInt64LE(zip64EocdOffset + 48));
        }
        break;
      }
    }
  }

  const entries = [];
  let pos = cdOffset;

  while (pos < cdOffset + cdSize && pos + 46 <= buf.length) {
    if (buf.readUInt32LE(pos) !== 0x02014b50) break;

    const method = buf.readUInt16LE(pos + 10);
    const compSize = buf.readUInt32LE(pos + 20);
    const uncompSize = buf.readUInt32LE(pos + 24);
    const nameLen = buf.readUInt16LE(pos + 28);
    const extraLen = buf.readUInt16LE(pos + 30);
    const commentLen = buf.readUInt16LE(pos + 32);
    const localOffset = buf.readUInt32LE(pos + 42);
    const name = buf.subarray(pos + 46, pos + 46 + nameLen).toString("utf8");

    entries.push({ name, method, compSize, uncompSize, localOffset });
    pos += 46 + nameLen + extraLen + commentLen;
  }

  return { entries, buf };
}

function extractZipEntry(zip, entry) {
  const { buf } = zip;
  const pos = entry.localOffset;
  if (buf.readUInt32LE(pos) !== 0x04034b50) return null;

  const nameLen = buf.readUInt16LE(pos + 26);
  const extraLen = buf.readUInt16LE(pos + 28);
  const dataStart = pos + 30 + nameLen + extraLen;
  const compressed = buf.subarray(dataStart, dataStart + entry.compSize);

  if (entry.method === 0) return compressed; // stored
  if (entry.method === 8) return inflateRawSync(compressed); // deflated
  return null; // unsupported method
}

// -- List ZIP entries --
function listZipEntries(zipPath) {
  try {
    const zip = readZipCentralDirectory(zipPath);
    return zip.entries || [];
  } catch {
    return [];
  }
}

// -- Extract single file from ZIP --
function unzipFile(zipPath, entryName) {
  try {
    const zip = readZipCentralDirectory(zipPath);
    const entry = zip.entries.find(e => e.name === entryName);
    if (!entry) return null;
    return extractZipEntry(zip, entry);
  } catch {
    return null;
  }
}

// -- Extract matching files from ZIP to temp directory --
function unzipFilesToDir(zipPath, pattern, destDir) {
  try {
    const zip = readZipCentralDirectory(zipPath);
    const re = new RegExp(pattern);
    const extracted = [];

    for (const entry of zip.entries) {
      if (!re.test(entry.name)) continue;
      const data = extractZipEntry(zip, entry);
      if (!data) continue;

      const dest = join(destDir, basename(entry.name));
      writeFileSync(dest, data);
      extracted.push(dest);
    }

    return extracted;
  } catch {
    return [];
  }
}

// -- Find all books (pairs of _uma.zip + _data.zip) --
function findBooks(basePath) {
  const books = [];
  for (const entry of readdirSync(basePath)) {
    if (entry.endsWith("_uma.zip")) {
      const id = entry.replace("_uma.zip", "");
      books.push({
        id,
        umaZip: join(basePath, entry),
        dataZip: existsSync(join(basePath, `${id}_data.zip`))
          ? join(basePath, `${id}_data.zip`)
          : null,
      });
    }
  }
  return books;
}

// -- Read uma.json from UMA zip --
function readUmaJson(umaZipPath) {
  const buf = unzipFile(umaZipPath, "uma.json");
  if (!buf) return null;
  try {
    return JSON.parse(buf.toString("utf8"));
  } catch {
    return null;
  }
}

// -- Find encrypted PDF entry in UMA zip --
function findPdfEntry(umaZipPath) {
  const entries = listZipEntries(umaZipPath);
  const pdf = entries.find(e => e.name.endsWith("_sf.pdf") || e.name.endsWith(".pdf"));
  return pdf ? pdf.name : null;
}

// -- Extract text via pdf.js --
async function extractPdfText(pdfPath) {
  if (!pdfjsLib) return null;
  try {
    const data = new Uint8Array(readFileSync(pdfPath));
    const doc = await pdfjsLib.getDocument({ data, useSystemFonts: true }).promise;
    const pages = [];

    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i);
      const content = await page.getTextContent();
      let text = "";
      for (const item of content.items) {
        text += item.str;
        if (item.hasEOL) text += "\n";
      }
      pages.push({ id: i, text: text.trim() });
    }

    await doc.destroy();
    return pages;
  } catch {
    return null;
  }
}

// -- Convert HTML to plain text --
function htmlToText(html) {
  return html
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<\/p>/gi, "\n\n")
    .replace(/<img[^>]*>/gi, "")
    .replace(/<[^>]+>/g, "")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&nbsp;/g, " ")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}


// -- Main --
async function main() {
  const args = process.argv.slice(2);
  const force = args.includes("--force");
  const markdown = args.includes("--markdown");
  const noMaterials = args.includes("--no-materials");
  const bookFilter = args.includes("--book")
    ? args[args.indexOf("--book") + 1]
    : null;
  const outputDir = args.includes("--output")
    ? resolve(args[args.indexOf("--output") + 1])
    : join(process.cwd(), "books");

  const BASE = findCornelsenBase();

  if (!existsSync(BASE)) {
    console.error("Cornelsen Offline Lernen Daten nicht gefunden unter:", BASE);
    console.error("Ist die Cornelsen App installiert und ein Buch heruntergeladen?");
    process.exit(1);
  }

  mkdirSync(outputDir, { recursive: true });

  const books = findBooks(BASE);
  if (books.length === 0) {
    console.error("Keine heruntergeladenen Bücher gefunden.");
    process.exit(1);
  }

  if (!pdfjsLib) {
    console.log("Hinweis: pdfjs-dist nicht installiert — nur PDF wird extrahiert, kein Text.");
    console.log("  npm install pdfjs-dist");
    console.log();
  }

  // Read metadata for all books
  console.log("Gefundene Bücher:");
  const umaCache = {};
  for (const book of books) {
    const uma = readUmaJson(book.umaZip);
    umaCache[book.id] = uma;
    console.log(`  ${uma?.title || book.id} (${book.id})`);
  }

  for (const book of books) {
    if (bookFilter && book.id !== bookFilter) continue;

    const uma = umaCache[book.id];
    const title = uma?.title || book.id;
    const isbn = uma?.isbnSb || "";

    console.log(`\n${title}:`);

    const baseName = title.replace(/[\/\\:*?"<>|]/g, "-");
    const dirName = isbn ? `${baseName} (${isbn})` : `${baseName} (${book.id})`;
    const bookDir = join(outputDir, dirName);

    const ext = markdown ? "md" : "txt";

    if (!force && existsSync(join(bookDir, `${baseName}.pdf`))) {
      console.log("  Bereits vorhanden, überspringe. (--force zum Überschreiben)");
      continue;
    }

    mkdirSync(bookDir, { recursive: true });

    // -- Decrypt and extract PDF --
    const pdfEntry = findPdfEntry(book.umaZip);

    if (pdfEntry) {
      console.log("  PDF entschlüsseln...");
      const encrypted = unzipFile(book.umaZip, pdfEntry);

      if (encrypted) {
        try {
          const decrypted = decryptBuffer(encrypted);
          const destPdf = join(bookDir, `${baseName}.pdf`);
          writeFileSync(destPdf, decrypted);
          const sizeMB = (decrypted.length / 1024 / 1024).toFixed(1);
          console.log(`  PDF: ${sizeMB} MB -> ${destPdf}`);

          if (pdfjsLib) {
            console.log("  Text extrahieren...");
            const pages = await extractPdfText(destPdf);

            if (pages) {
              const nonEmpty = pages.filter(p => p.text);
              console.log(`  Text: ${nonEmpty.length}/${pages.length} Seiten`);

              const outPath = join(bookDir, `${baseName}.${ext}`);
              let content = "";

              if (markdown) {
                content = `# ${title}\n\n`;
                for (const p of pages) {
                  if (!p.text) continue;
                  content += `## Seite ${p.id}\n\n${p.text}\n\n`;
                }
              } else {
                for (const p of pages) {
                  content += `--- Seite ${p.id} ---\n${p.text}\n\n`;
                }
              }

              writeFileSync(outPath, content, "utf8");
              console.log(`  -> ${outPath}`);
            }
          }
        } catch (e) {
          console.error(`  Entschlüsselung fehlgeschlagen: ${e.message}`);
        }
      }
    } else {
      console.log("  Kein PDF im UMA-Archiv gefunden.");
    }

    // -- Export supplementary materials --
    if (!noMaterials && book.dataZip) {
      const entries = listZipEntries(book.dataZip);
      const htmlEntries = entries.filter(e => e.name.endsWith(".html"));
      const pdfEntries = entries.filter(e => e.name.endsWith(".pdf"));

      const totalMat = htmlEntries.length + pdfEntries.length;

      if (totalMat > 0) {
        console.log(`  Materialien: ${htmlEntries.length} Tipps, ${pdfEntries.length} PDFs`);
        const matDir = join(bookDir, "Zusatzmaterial");
        mkdirSync(matDir, { recursive: true });

        // Extract HTML tips into single Tipps.md
        if (htmlEntries.length > 0) {
          // Build elvisId -> asset metadata map from uma.json
          const assetMap = {};
          if (uma?.assets) {
            for (const a of uma.assets) {
              const key = a.elvisId || basename(a.fileName || "", ".html");
              if (key) assetMap[key] = a;
            }
          }

          try {
            const zip = readZipCentralDirectory(book.dataZip);

            // Sort by page number from metadata, then by filename
            const sorted = htmlEntries.slice().sort((a, b) => {
              const idA = basename(a.name, ".html");
              const idB = basename(b.name, ".html");
              const pageA = parseInt(assetMap[idA]?.pageFrom) || 9999;
              const pageB = parseInt(assetMap[idB]?.pageFrom) || 9999;
              return pageA - pageB || idA.localeCompare(idB);
            });

            let md = `# Tipps — ${title}\n\n`;
            md += `${sorted.length} Aufgaben-Tipps.\n\n`;
            let converted = 0;

            for (const entry of sorted) {
              const zipEntry = zip.entries.find(e => e.name === entry.name);
              if (!zipEntry) continue;
              const buf = extractZipEntry(zip, zipEntry);
              if (!buf) continue;

              const html = buf.toString("utf8");
              const text = htmlToText(html);
              if (!text) continue;

              const elvisId = basename(entry.name, ".html");
              const meta = assetMap[elvisId];
              const page = meta?.pageFrom ? `Seite ${meta.pageFrom}` : "";
              const sub = meta?.subtitleHere || "";
              const heading = [page, sub].filter(Boolean).join(" — ") || elvisId;

              md += `## ${heading}\n\n${text}\n\n---\n\n`;
              converted++;
            }

            writeFileSync(join(matDir, "Tipps.md"), md, "utf8");
            console.log(`  -> ${converted} Tipps in Tipps.md zusammengefasst`);
          } catch (e) {
            console.error(`  Tipp-Extraktion fehlgeschlagen: ${e.message}`);
          }
        }

        // Extract non-encrypted PDFs from data zip and convert to markdown
        if (pdfEntries.length > 0) {
          let converted = 0;

          for (const entry of pdfEntries) {
            const buf = unzipFile(book.dataZip, entry.name);
            if (!buf || buf[0] !== 0x25) continue; // not %PDF

            const name = basename(entry.name, ".pdf");
            const tmpPath = join(tmpdir(), `cornelsen_mat_${name}.pdf`);
            writeFileSync(tmpPath, buf);

            const pages = await extractPdfText(tmpPath);
            try { rmSync(tmpPath, { force: true }); } catch {}

            if (pages) {
              const nonEmpty = pages.filter(p => p.text);
              if (nonEmpty.length > 0) {
                let md = `# ${name}\n\n`;
                for (const p of nonEmpty) {
                  md += `${p.text}\n\n`;
                }
                writeFileSync(join(matDir, `${name}.md`), md, "utf8");
                converted++;
              }
            }
          }

          if (converted > 0) console.log(`  -> ${converted} Material-PDFs nach Markdown konvertiert`);
        }

        // Material index from uma.json assets
        if (uma?.assets) {
          let md = `# Zusatzmaterial — ${title}\n\n`;
          md += `Insgesamt ${uma.assets.length} Assets.\n\n`;

          const byType = {};
          for (const a of uma.assets) {
            const type = a.assetType || "sonstige";
            (byType[type] ??= []).push(a);
          }

          for (const [type, items] of Object.entries(byType).sort()) {
            md += `## ${type} (${items.length})\n\n`;
            for (const a of items) {
              md += `- ${a.title || a.fileName || a.id}\n`;
            }
            md += "\n";
          }

          writeFileSync(join(bookDir, "Zusatzmaterial.md"), md, "utf8");
        }
      } else {
        console.log("  Kein Zusatzmaterial gefunden.");
      }
    }
  }

  console.log("\nFertig.");
}

main();

File: cornelsen.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = ["pymupdf", "cryptography"]
# ///
"""
Extract books from Cornelsen Offline Lernen data.

Reads the Cornelsen app container, decrypts the embedded PDF (AES-128-CBC),
extracts page text via PyMuPDF, and exports supplementary materials
(HTML tips, additional PDFs).

Usage: cornelsen [--output <dir>] [--book <id>]
                 [--no-materials] [--markdown] [--force]
"""

import sys
import os
import re
import json
import struct
import zlib
import tempfile
from pathlib import Path
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

# -- Try to import PyMuPDF for text extraction --
try:
    import fitz  # PyMuPDF
    HAS_FITZ = True
except ImportError:
    HAS_FITZ = False

# -- AES-128-CBC decryption (key from app source) --
CIPHER_B64 = "YWVzLTEyOC1jYmN8RCtEeEpTRn0yQjtrLTtDfQ=="

def decrypt_buffer(encrypted: bytes) -> bytes:
    import base64
    decoded = base64.b64decode(CIPHER_B64).decode("ascii")
    algorithm, key_str = decoded.split("|")
    key = key_str.encode("ascii")
    iv = key_str[::-1].encode("ascii")

    from cryptography.hazmat.primitives import padding as sym_padding
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
    decryptor = cipher.decryptor()
    padded = decryptor.update(encrypted) + decryptor.finalize()

    # Remove PKCS7 padding
    unpadder = sym_padding.PKCS7(128).unpadder()
    return unpadder.update(padded) + unpadder.finalize()


# -- Path to Cornelsen app data --
def find_cornelsen_base() -> Path:
    home = Path.home()
    candidates = []

    if sys.platform == "darwin":
        candidates = [
            Path("/Applications/CornelsenOfflineLernen.app/Contents/Resources/uma"),
            home / "Applications" / "CornelsenOfflineLernen.app" / "Contents" / "Resources" / "uma",
        ]
    elif sys.platform == "win32":
        local = Path(os.environ.get("LOCALAPPDATA", home / "AppData" / "Local"))
        prog = Path(os.environ.get("PROGRAMFILES", "C:\\Program Files"))
        prog86 = Path(os.environ.get("PROGRAMFILES(X86)", "C:\\Program Files (x86)"))

        for base in (local / "Programs", prog, prog86):
            candidates.append(base / "CornelsenOfflineLernen" / "resources" / "uma")
            candidates.append(base / "CornelsenOfflineLernen" / "uma")
            candidates.append(base / "Cornelsen Offline Lernen" / "resources" / "uma")
            candidates.append(base / "Cornelsen Offline Lernen" / "uma")

    for p in candidates:
        if p.exists():
            return p
    return candidates[0] if candidates else Path("cornelsen_uma")


# -- Pure Python ZIP reader --

def read_zip_central_directory(zip_path: Path) -> tuple[list[dict], bytes]:
    buf = zip_path.read_bytes()

    # Find End of Central Directory
    eocd_offset = -1
    for i in range(len(buf) - 22, max(-1, len(buf) - 65557 - 1), -1):
        if struct.unpack_from("<I", buf, i)[0] == 0x06054B50:
            eocd_offset = i
            break
    if eocd_offset < 0:
        return [], buf

    cd_offset = struct.unpack_from("<I", buf, eocd_offset + 16)[0]
    cd_size = struct.unpack_from("<I", buf, eocd_offset + 12)[0]

    # ZIP64
    if cd_offset == 0xFFFFFFFF:
        for i in range(eocd_offset - 20, max(-1, eocd_offset - 41), -1):
            if struct.unpack_from("<I", buf, i)[0] == 0x07064B50:
                z64_offset = struct.unpack_from("<Q", buf, i + 8)[0]
                if struct.unpack_from("<I", buf, z64_offset)[0] == 0x06064B50:
                    cd_offset = struct.unpack_from("<Q", buf, z64_offset + 48)[0]
                break

    entries = []
    pos = cd_offset
    while pos < cd_offset + cd_size and pos + 46 <= len(buf):
        sig = struct.unpack_from("<I", buf, pos)[0]
        if sig != 0x02014B50:
            break
        method = struct.unpack_from("<H", buf, pos + 10)[0]
        comp_size = struct.unpack_from("<I", buf, pos + 20)[0]
        uncomp_size = struct.unpack_from("<I", buf, pos + 24)[0]
        name_len = struct.unpack_from("<H", buf, pos + 28)[0]
        extra_len = struct.unpack_from("<H", buf, pos + 30)[0]
        comment_len = struct.unpack_from("<H", buf, pos + 32)[0]
        local_offset = struct.unpack_from("<I", buf, pos + 42)[0]
        name = buf[pos + 46: pos + 46 + name_len].decode("utf-8")

        entries.append({
            "name": name, "method": method,
            "comp_size": comp_size, "uncomp_size": uncomp_size,
            "local_offset": local_offset,
        })
        pos += 46 + name_len + extra_len + comment_len

    return entries, buf


def extract_zip_entry(buf: bytes, entry: dict) -> bytes | None:
    pos = entry["local_offset"]
    sig = struct.unpack_from("<I", buf, pos)[0]
    if sig != 0x04034B50:
        return None
    name_len = struct.unpack_from("<H", buf, pos + 26)[0]
    extra_len = struct.unpack_from("<H", buf, pos + 28)[0]
    data_start = pos + 30 + name_len + extra_len
    compressed = buf[data_start: data_start + entry["comp_size"]]

    if entry["method"] == 0:
        return compressed
    if entry["method"] == 8:
        return zlib.decompress(compressed, -15)
    return None


def unzip_file(zip_path: Path, entry_name: str) -> bytes | None:
    try:
        entries, buf = read_zip_central_directory(zip_path)
        entry = next((e for e in entries if e["name"] == entry_name), None)
        if not entry:
            return None
        return extract_zip_entry(buf, entry)
    except Exception:
        return None


def list_zip_entries(zip_path: Path) -> list[dict]:
    try:
        entries, _ = read_zip_central_directory(zip_path)
        return entries
    except Exception:
        return []


# -- Find all books --
def find_books(base_path: Path) -> list[dict]:
    books = []
    for entry in sorted(base_path.iterdir()):
        if entry.name.endswith("_uma.zip"):
            book_id = entry.name.replace("_uma.zip", "")
            data_zip = base_path / f"{book_id}_data.zip"
            books.append({
                "id": book_id,
                "uma_zip": entry,
                "data_zip": data_zip if data_zip.exists() else None,
            })
    return books


# -- Read uma.json from UMA zip --
def read_uma_json(uma_zip: Path) -> dict | None:
    buf = unzip_file(uma_zip, "uma.json")
    if not buf:
        return None
    try:
        return json.loads(buf.decode("utf-8"))
    except Exception:
        return None


# -- Find encrypted PDF entry --
def find_pdf_entry(uma_zip: Path) -> str | None:
    entries = list_zip_entries(uma_zip)
    for e in entries:
        if e["name"].endswith("_sf.pdf") or e["name"].endswith(".pdf"):
            return e["name"]
    return None


# -- Extract text via PyMuPDF --
def extract_pdf_text(pdf_path: Path) -> list[dict] | None:
    if not HAS_FITZ:
        return None
    try:
        doc = fitz.open(str(pdf_path))
        pages = []
        for i in range(len(doc)):
            page = doc[i]
            text = page.get_text().strip()
            pages.append({"id": i + 1, "text": text})
        doc.close()
        return pages
    except Exception:
        return None


# -- Convert HTML to plain text --
def html_to_text(html: str) -> str:
    text = re.sub(r"<br\s*/?>", "\n", html, flags=re.I)
    text = re.sub(r"</p>", "\n\n", text, flags=re.I)
    text = re.sub(r"<img[^>]*>", "", text, flags=re.I)
    text = re.sub(r"<[^>]+>", "", text)
    text = text.replace("&amp;", "&").replace("&lt;", "<")
    text = text.replace("&gt;", ">").replace("&quot;", '"').replace("&nbsp;", " ")
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()


# -- Main --
def main():
    args = sys.argv[1:]
    force = "--force" in args
    markdown = "--markdown" in args
    no_materials = "--no-materials" in args
    book_filter = args[args.index("--book") + 1] if "--book" in args else None
    default_books = Path(__file__).resolve().parent / "books"
    output_dir = Path(args[args.index("--output") + 1]).resolve() if "--output" in args else default_books

    base = find_cornelsen_base()

    if not base.exists():
        print(f"Cornelsen Offline Lernen Daten nicht gefunden unter: {base}", file=sys.stderr)
        print("Ist die Cornelsen App installiert und ein Buch heruntergeladen?", file=sys.stderr)
        sys.exit(1)

    output_dir.mkdir(parents=True, exist_ok=True)

    books = find_books(base)
    if not books:
        print("Keine heruntergeladenen Bücher gefunden.", file=sys.stderr)
        sys.exit(1)

    if not HAS_FITZ:
        print("Hinweis: PyMuPDF nicht verfügbar — nur PDF wird extrahiert, kein Text.")
        print()

    # Read metadata
    print("Gefundene Bücher:")
    uma_cache = {}
    for book in books:
        uma = read_uma_json(book["uma_zip"])
        uma_cache[book["id"]] = uma
        print(f"  {uma.get('title', book['id']) if uma else book['id']} ({book['id']})")

    for book in books:
        if book_filter and book["id"] != book_filter:
            continue

        uma = uma_cache.get(book["id"])
        title = uma.get("title", book["id"]) if uma else book["id"]
        isbn = uma.get("isbnSb", "") if uma else ""

        print(f"\n{title}:")

        base_name = re.sub(r'[/\\:*?"<>|]', "-", title)
        dir_name = f"{base_name} ({isbn})" if isbn else f"{base_name} ({book['id']})"
        book_dir = output_dir / dir_name

        if not force and (book_dir / f"{base_name}.pdf").exists():
            print("  Bereits vorhanden, überspringe. (--force zum Überschreiben)")
            continue

        book_dir.mkdir(parents=True, exist_ok=True)

        # Decrypt and extract PDF
        pdf_entry = find_pdf_entry(book["uma_zip"])

        if pdf_entry:
            print("  PDF entschlüsseln...")
            encrypted = unzip_file(book["uma_zip"], pdf_entry)

            if encrypted:
                try:
                    decrypted = decrypt_buffer(encrypted)
                    dest_pdf = book_dir / f"{base_name}.pdf"
                    dest_pdf.write_bytes(decrypted)
                    size_mb = len(decrypted) / 1024 / 1024
                    print(f"  PDF: {size_mb:.1f} MB -> {dest_pdf}")

                    if HAS_FITZ:
                        print("  Text extrahieren...")
                        pages = extract_pdf_text(dest_pdf)

                        if pages:
                            non_empty = [p for p in pages if p["text"]]
                            print(f"  Text: {len(non_empty)}/{len(pages)} Seiten")

                            ext = "md" if markdown else "txt"
                            out_path = book_dir / f"{base_name}.{ext}"

                            if markdown:
                                content = f"# {title}\n\n"
                                for p in pages:
                                    if not p["text"]:
                                        continue
                                    content += f"## Seite {p['id']}\n\n{p['text']}\n\n"
                            else:
                                content = ""
                                for p in pages:
                                    content += f"--- Seite {p['id']} ---\n{p['text']}\n\n"

                            out_path.write_text(content, encoding="utf-8")
                            print(f"  -> {out_path}")
                except Exception as e:
                    print(f"  Entschlüsselung fehlgeschlagen: {e}", file=sys.stderr)
        else:
            print("  Kein PDF im UMA-Archiv gefunden.")

        # Export supplementary materials
        if not no_materials and book.get("data_zip"):
            entries = list_zip_entries(book["data_zip"])
            html_entries = [e for e in entries if e["name"].endswith(".html")]
            pdf_entries = [e for e in entries if e["name"].endswith(".pdf")]

            total_mat = len(html_entries) + len(pdf_entries)

            if total_mat > 0:
                print(f"  Materialien: {len(html_entries)} Tipps, {len(pdf_entries)} PDFs")
                mat_dir = book_dir / "Zusatzmaterial"
                mat_dir.mkdir(parents=True, exist_ok=True)

                # Build asset map from uma.json
                asset_map = {}
                if uma and uma.get("assets"):
                    for a in uma["assets"]:
                        key = a.get("elvisId") or Path(a.get("fileName", "")).stem
                        if key:
                            asset_map[key] = a

                # Extract HTML tips into single Tipps.md
                if html_entries:
                    try:
                        all_entries, buf = read_zip_central_directory(book["data_zip"])

                        sorted_html = sorted(html_entries, key=lambda e: (
                            int(asset_map.get(Path(e["name"]).stem, {}).get("pageFrom", 9999) or 9999),
                            Path(e["name"]).stem,
                        ))

                        md = f"# Tipps — {title}\n\n"
                        md += f"{len(sorted_html)} Aufgaben-Tipps.\n\n"
                        converted = 0

                        for entry in sorted_html:
                            zip_entry = next((e for e in all_entries if e["name"] == entry["name"]), None)
                            if not zip_entry:
                                continue
                            data = extract_zip_entry(buf, zip_entry)
                            if not data:
                                continue

                            html = data.decode("utf-8")
                            text = html_to_text(html)
                            if not text:
                                continue

                            elvis_id = Path(entry["name"]).stem
                            meta = asset_map.get(elvis_id, {})
                            page = f"Seite {meta['pageFrom']}" if meta.get("pageFrom") else ""
                            sub = meta.get("subtitleHere", "")
                            heading = " — ".join(filter(None, [page, sub])) or elvis_id

                            md += f"## {heading}\n\n{text}\n\n---\n\n"
                            converted += 1

                        (mat_dir / "Tipps.md").write_text(md, encoding="utf-8")
                        print(f"  -> {converted} Tipps in Tipps.md zusammengefasst")
                    except Exception as e:
                        print(f"  Tipp-Extraktion fehlgeschlagen: {e}", file=sys.stderr)

                # Extract PDFs and convert to markdown
                if pdf_entries and HAS_FITZ:
                    converted = 0
                    for entry in pdf_entries:
                        data = unzip_file(book["data_zip"], entry["name"])
                        if not data or data[0:1] != b"%":
                            continue

                        name = Path(entry["name"]).stem
                        tmp_path = Path(tempfile.gettempdir()) / f"cornelsen_mat_{name}.pdf"
                        tmp_path.write_bytes(data)

                        pages = extract_pdf_text(tmp_path)
                        try:
                            tmp_path.unlink(missing_ok=True)
                        except Exception:
                            pass

                        if pages:
                            non_empty = [p for p in pages if p["text"]]
                            if non_empty:
                                md = f"# {name}\n\n"
                                for p in non_empty:
                                    md += f"{p['text']}\n\n"
                                (mat_dir / f"{name}.md").write_text(md, encoding="utf-8")
                                converted += 1

                    if converted > 0:
                        print(f"  -> {converted} Material-PDFs nach Markdown konvertiert")

                # Material index
                if uma and uma.get("assets"):
                    md = f"# Zusatzmaterial — {title}\n\n"
                    md += f"Insgesamt {len(uma['assets'])} Assets.\n\n"

                    by_type: dict[str, list] = {}
                    for a in uma["assets"]:
                        t = a.get("assetType", "sonstige")
                        by_type.setdefault(t, []).append(a)

                    for t, items in sorted(by_type.items()):
                        md += f"## {t} ({len(items)})\n\n"
                        for a in items:
                            md += f"- {a.get('title') or a.get('fileName') or a.get('id')}\n"
                        md += "\n"

                    (book_dir / "Zusatzmaterial.md").write_text(md, encoding="utf-8")
            else:
                print("  Kein Zusatzmaterial gefunden.")

    print("\nFertig.")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"\nFehler: {e}", file=sys.stderr, flush=True)
    if getattr(sys, "frozen", False):
        input("\nDrücke Enter zum Beenden...")
